firmware-support/bittide-sys/src/dna_port_e2.rs

```diff
@@ -0,0 +1,33 @@
+// SPDX-FileCopyrightText: 2024 Google LLC
+//
+// SPDX-License-Identifier: Apache-2.0
+/*! DNA Port E2 peripheral interface.
+
+This module contains an interface to read the DNA port E2 peripheral.
+The DNA value is a 96-bit unique identifier per device, the interface to device
+identifier is offered by the `DnaPortE2` struct.
+*/
+
+/// The DnaPortE2 struct is a hardware abstraction for the DnaPortE2 peripheral,
+/// a unqique identifier for the FPGA device.
+pub struct DnaPortE2 {}
+
+impl DnaPortE2 {
+    /// Reads the DNA value from the DNA port E2 peripheral.
+    /// The DNA value is a 96-bit unique identifier, the upper bits from the 128-bit value are zero.
+    pub fn read(addr: usize) -> u128 {
+        let mut tmp;
+        unsafe {
+            tmp = (addr as *const u32).add(2);
+        }
+        let mut result = 0;
+        for _ in 0..3 {
+            unsafe {
+                result <<= 32;
+                result |= tmp.read_volatile() as u128;
+                tmp = tmp.sub(1);
+            }
+        }
+        result
+    }
```

Comment:
```suggestion
    /// Reads the DNA value from the DNA port E2 peripheral.
    /// The DNA value is a 96-bit unique identifier, the upper bits from the 128-bit value are zero.
    ///
    /// # Safety
    /// The `addr` pointer must point to valid memory field of at least size 12 bytes.
    pub unsafe fn read(addr: *const u32) -> u128 {
        let mut tmp = addr.add(2);
  
        let mut result = 0;
        for _ in 0..3 {
            result <<= 32;
            result |= tmp.read_volatile() as u128;
            tmp = tmp.sub(1);
        }
        result
    }
```

This function cannot be marked as safe, it must be `unsafe` because
- an arbitrary value is passed in, that's interpreted as a pointer
- that pointer gets dereferenced without any safety checks

So it's really easy to cause undefined behaviour with *safe Rust*, which is the whole point that safe Rust is trying to prevent.

(It would be nice if we could work more with array here instead of `u128` and loops and shifts, but Rust's const generics aren't there yet for that :pensive: )

Reply:
I'm also not very happy with this code tbh. I did look at different solutions and found some packages that could represent `u96`, but I myself found that a bit much to add new dependencies just to read a single unsigned number...